### PR TITLE
Fix issue where multiple dist files were not being produced for comma separated formats value

### LINKED ISSUE #8655: Multiple dist formats doesn't work as stated in help documentation
Seems that `meson dist -C build --formats=xztar,gztar` no longer works. Help documentation for `--formats` says: `Comma separated list of archive types to create.`

### PR DESCRIPTION
**Pull request by tristan957**
Fixes #8655

### CODE REVIEW

**Comment by tristan957** [comment]
Took away the choices because the help text looks nasty listing all the permutations. This restores the behavior from what looks like a year ago if Git blame is right.

File: mesonbuild/mdist.py [hunk 1]
```diff
@@ -29,14 +31,25 @@
...
+def __archive_permutations() -> T.Tuple[str]:
+    permutations: T.List[str] = []
+    for i in range(len(archive_choices) + 1):
+        for subset in itertools.permutations(archive_choices, i):
+            if len(subset) > 0:
+                permutations.append(','.join(subset))
+
+    return tuple(permutations)
+
+archive_permutations = __archive_permutations()
```
Comment (xclaesse) [inline mesonbuild/mdist.py:44]: `archive_permutations` does not seems to be used anywhere.

File: mesonbuild/mdist.py [hunk 2]
```diff
@@ -29,14 +31,25 @@
...
 archive_extension = {'gztar': '.tar.gz',
                      'xztar': '.tar.xz',
                      'zip': '.zip'}
 
 def add_arguments(parser):
     parser.add_argument('-C', default='.', dest='wd',
                         help='directory to cd into before running')
-    parser.add_argument('--formats', default='xztar', choices=archive_choices,
+    parser.add_argument('--formats', default='xztar',
                         help='Comma separated list of archive types to create.')
```
Comment (xclaesse) [inline mesonbuild/mdist.py:53]: You can add list of supported archives in the help message.

File: run_unittests.py [hunk 3]
```diff
@@ -3232,15 +3235,52 @@ def dist_impl(self, vcs_init, vcs_add_all=None, include_subprojects=True):
...
-            self.assertPathDoesNotExist(zip_checksumfile)
-            self._run(self.meson_command + ['dist', '--formats', 'zip'],
-                      workdir=self.builddir)
-            self.assertPathExists(zip_distfile)
-            self.assertPathExists(zip_checksumfile)
+            self.build()
+
+            for i in range(len(archive_choices) + 1):
+                permutations = tuple(itertools.permutations(archive_choices, i))
+                for i, subset in enumerate(permutations):
```
Comment (xclaesse) [inline run_unittests.py:3242]: Is it really worth testing all possible permutations? I would just test the case where you have all archives with `','.join(archive_choices)`.

File: mesonbuild/mdist.py (hunk 2 above)
Comment (tristan957) [reply mesonbuild/mdist.py:53]: Meant to do that.

File: run_unittests.py (hunk 3 above)
Comment (tristan957) [reply run_unittests.py:3242]: Good point. I think will do as you have said.

File: run_unittests.py [hunk 4]
```diff
@@ -3232,15 +3235,52 @@ def dist_impl(self, vcs_init, vcs_add_all=None, include_subprojects=True):
...
+                    # skip empty subset
+                    if len(subset) == 0:
```
Comment (dcbaker) [inline run_unittests.py:3244]: using `len()` to test for container emptiness is a python anti-pattern, just:
```python
if not subset:
```

File: run_unittests.py (hunk 3 above)
Comment (dcbaker) [reply run_unittests.py:3242]: At least one test that isn't all of the options is probably useful. I've been bitten enough times by "but I tested x + y + z and it worked, why doesn't just x wor‽"

File: run_unittests.py (hunk 4 above)
Comment (tristan957) [reply run_unittests.py:3244]: Good to know.

File: run_unittests.py [hunk 5]
```diff
@@ -3235,12 +3237,20 @@ def dist_impl(self, vcs_init, vcs_add_all=None, include_subprojects=True):
...
+            self._run(self.meson_command + ['dist', '--formats', 'gztar'],
+                      workdir=self.builddir)
+            self.assertPathExists(gz_distfile)
+            self.assertPathExists(gz_checksumfile)
             self._run(self.meson_command + ['dist', '--formats', 'zip'],
                       workdir=self.builddir)
             self.assertPathExists(zip_distfile)
             self.assertPathExists(zip_checksumfile)
+            self._run(self.meson_command + ['dist', '--formats', 'xztar,gztar,zip'],
+                      workdir=self.builddir)
```
Comment (jpakkane) [inline run_unittests.py:3253]: This should check that the output files did get generated.

**Comment by jpakkane** [comment]
Test failure is not related.

**Pull request merged**